### PR TITLE
Enable GSI interrupt on virtio devices

### PR DESCRIPTION
**Pull request by michael2012z**
The PR enables the support of GSI interrupt on Virtio devices.

Now this is a RFC to evaluate. Please feel free to comment.

The intension of adding this support is to enable Cloud Hypervisor on broader range of Arm64 development boards. Now our virtio-pci device requires MSI-X. On AArch64 architecture, MSI-X requires GICv3 hardware on host. But GICv3 is not available on many development boards (among them is Raspberry Pi 4B).

Summary of the PR:
- Allocate a GSI and create the legacy interrupt source group for each VirtioPciDevice.
- Determine the type of interrupt should be used (MSI-X or GSI) after device initialization. If the driver didn't query the MSI-X capability, GSI will be used. By default guest use MSI-X. Applying `pci=nomsi` in guest kernel command line can change.
- To support level-triggered interrupt, use `resamplefd` together with `irqfd` to receive the `de-assertion` of level triggered interrupt from guest.
- To poll the `resamplefd` event, add a epoll handler for common purpose of all sorts of virtio devices.
- AArch64 FDT update for interrupt mapping.

Somethings to notice:
- To support the `resamplefd`, a new API is required in `rust-vmm/kvm-ioctl`. Now I am testing with a private `kvm-ioctl` repo. Will upload the patch to `rust-vmm` if this proposal is agreed.
- Haven't done enough test on all types of virtio devices and on all kinds of scenarios. There should be gaps in migration, vhost-users, vfio and etc.
- The GSI solution itself should have some limitations, for example the maximum number of devices.
- I also have some questions in implementing the proposal, I will add comments to discuss.
- I have tested only AArch64 so far, not x86.

### CODE REVIEW

File: virtio-devices/src/net.rs [hunk 1]
```diff
@@ -587,6 +587,7 @@ impl VirtioDevice for Net {
         interrupt_cb: Arc<dyn VirtioInterrupt>,
         mut queues: Vec<Queue<GuestMemoryAtomic<GuestMemoryMmap>>>,
         mut queue_evts: Vec<EventFd>,
+        _resample_evt: Option<EventFd>,
```
Comment (michael2012z) [inline virtio-devices/src/net.rs:590]: Not all devices need resampling. In my tests, resampling was only seen in the first interrupt of virtio-blk device. So I am confused on what type of interrupt level the devices are using (level-triggerd or edge-triggered). 

Please comment if you know about this.

File: virtio-devices/src/transport/pci_device.rs [hunk 2]
```diff
@@ -792,6 +829,84 @@ impl VirtioInterrupt for VirtioInterruptMsix {
...
+    }
+}
+
+impl VirtioInterrupt for VirtioInterruptLegacy {
+    fn enable(&self) -> std::result::Result<(), std::io::Error> {
+        self.legacy_interrupt_source_group.enable().unwrap();
+        let config = LegacyIrqSourceConfig {
+            irqchip: 0,
+            pin: (self.gsi) as u32,
+        };
```
Comment (michael2012z) [inline virtio-devices/src/transport/pci_device.rs:864]: I am not sure whether this setting work on x86 or not.

File: virtio-devices/src/net.rs (hunk 1 above)
Comment (michael2012z) [reply virtio-devices/src/net.rs:590]: For your reference:
https://patchwork.kernel.org/project/kvm/patch/20120921175601.32604.63271.stgit@bling.home/

**Comment by sboeuf** [comment]
@michael2012z 
>The intension of adding this support is to enable Cloud Hypervisor on broader range of Arm64 development boards. Now our virtio-pci device requires MSI-X. On AArch64 architecture, MSI-X requires GICv3 hardware on host. But GICv3 is not available on many development boards (among them is Raspberry Pi 4B).

This PR is going a bit against the project's motto. We're increasing the complexity of the project (both code maintenance and CI) to support things like a Raspberry Pi.
The goal is to support modern cloud workloads on modern cloud hardware. I agree a RaspberryPi 4 can be considered as modern, but it can't really be considered as cloud though (more like IOT).

**Comment by michael2012z** [comment]
Cancel this PR. May come back to the topic if more requirements for legacy IRQ emerge in future.
Thanks for discussing.

**Pull request closed without merging**